### PR TITLE
feat: k8s resource

### PR DESCRIPTION
**Pull request by Sh4d1**
Signed-off-by: Patrik Cyvoct <patrik@ptrk.io>

### CODE REVIEW

File: scaleway/resource_k8s_pool.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+			},
+			"container_runtime": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "Container runtime for the pool",
+			},
+			"region":          regionSchema(),
+			"organization_id": organizationIDSchema(),
+			// Computed elements
+			"id": {
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_pool.go]: id should not be included in the schema as it's required for every resource

File: go.mod [hunk 2]
```diff
@@ -1,13 +1,18 @@
 module github.com/terraform-providers/terraform-provider-scaleway
```
[the remark this answers is not in the document]
Comment (Sh4d1) [reply go.mod]: Different commit or different PR?

[the remark this answers is not in the document]
Comment (kindermoumoute) [reply go.mod]: Different PR if possible

File: scaleway/helpers_k8s.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+func getK8SAPIWithRegionAndID(m interface{}, id string) (*k8s.API, scw.Region, string, error) {
+	meta := m.(*Meta)
+	k8sAPI := k8s.NewAPI(meta.scwClient)
+
+	region, ID, err := parseRegionalID(id)
+	return k8sAPI, region, ID, err
+}
+
+func waitK8SClusterReady(k8sAPI *k8s.API, region scw.Region, clusterID string) error {
```
Comment (kindermoumoute) [inline scaleway/helpers_k8s.go]: this method should live in the SDK

File: scaleway/helpers_k8s.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	for {
+		response, err := k8sAPI.GetCluster(&k8s.GetClusterRequest{
+			Region:    region,
+			ClusterID: clusterID,
+		})
+		if err != nil {
+			return err
+		}
+
+		if response.Status == k8s.ClusterStatusReady {
```
Comment (kindermoumoute) [inline scaleway/helpers_k8s.go]: Should handle all final states (unknown, ready, warning, error, locked)

File: scaleway/helpers_k8s.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+			return nil
+		}
+		time.Sleep(time.Second * 5)
```
Comment (kindermoumoute) [inline scaleway/helpers_k8s.go]: When moving this method to the SDK, use [async.WaitSync](https://github.com/scaleway/scaleway-sdk-go/blob/master/api/instance/v1/server_utils.go#L67)

File: scaleway/helpers_k8s.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		if time.Now().After(now.Add(K8SClusterWaitForReadyTimeout)) {
+			break
+		}
+	}
+
+	return fmt.Errorf("Timeout waiting for cluster %s to be ready", clusterID)
+}
+
+func waitK8SClusterDeleted(k8sAPI *k8s.API, region scw.Region, clusterID string) error {
```
Comment (kindermoumoute) [inline scaleway/helpers_k8s.go]: this method should live in the SDK

File: scaleway/resource_k8s_cluster_beta.go [hunk 7]
```diff
@@ -0,0 +1,733 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				ForceNew:    true,
+				Description: "The name of the cluster",
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Can't the description be empty? Why is it computed? 🤔

File: scaleway/resource_k8s_cluster_beta.go [hunk 8]
```diff
@@ -0,0 +1,733 @@
...
+				Description: "The description of the cluster",
+			},
+			"version": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Optional:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: remove line

File: scaleway/resource_k8s_cluster_beta.go [hunk 9]
```diff
@@ -0,0 +1,733 @@
...
+				Required:    true,
+				ForceNew:    true,
+				Description: "The CNI plugin of the cluster",
+			},
+			"enable_dashboard": {
+				Type:        schema.TypeBool,
+				ForceNew:    true,
+				Optional:    true,
+				Computed:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set the default value here (true or false), and remove the `Computed`.

File: scaleway/resource_k8s_cluster_beta.go [hunk 10]
```diff
@@ -0,0 +1,733 @@
...
+				Description: "Enable the dashboard on the cluster",
+			},
+			"ingress": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Computed:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set the default value here, and remove the Computed.

File: scaleway/resource_k8s_cluster_beta.go [hunk 11]
```diff
@@ -0,0 +1,733 @@
...
+				MaxItems:    1,
+				Computed:    true,
+				Optional:    true,
+				Description: "Default pool created on the cluster on creation",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"node_type": {
+							Type:        schema.TypeString,
+							Required:    true,
+							ForceNew:    true,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: note: add a test to make sure changing the node type recreate the cluster.

File: scaleway/resource_k8s_cluster_beta.go [hunk 12]
```diff
@@ -0,0 +1,733 @@
...
+							Description: "Server Type of the default pool servers",
+						},
+						"autoscaling": {
+							Type:        schema.TypeBool,
+							Optional:    true,
+							Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set the default value here (true or false), and remove the Computed.  (might not work in a list though)

File: scaleway/resource_k8s_cluster_beta.go [hunk 13]
```diff
@@ -0,0 +1,733 @@
...
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     1,
+							Description: "Minimun size of the default pool",
+						},
+						"max_size": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Computed:    true,
+							Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: remove this line

File: scaleway/resource_k8s_cluster_beta.go [hunk 14]
```diff
@@ -0,0 +1,733 @@
...
+							Description: "Maximum size of the default pool",
+						},
+						"container_runtime": {
+							Type:        schema.TypeString,
+							Optional:    true,
+							Computed:    true,
+							ForceNew:    true,
+							Description: "Container runtime for the default pool",
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set the default value, and remove the Computed. (might not work in a list though)

File: scaleway/resource_k8s_cluster_beta.go [hunk 15]
```diff
@@ -0,0 +1,733 @@
...
+			}
+			defaultPool := map[string]interface{}{}
+			defaultPool["pool_id"] = defaultPoolRes.ID
+
+			d.Set("default_pool", []map[string]interface{}{defaultPool})
+
+		}
+	}
+
+	err = resourceScalewayK8SClusterBetaDefaultPoolRead(d, m)
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Duplicate lines

File: scaleway/resource_k8s_cluster_beta.go [hunk 16]
```diff
@@ -0,0 +1,733 @@
...
+	d.Set("cni", response.Cni)
+	d.Set("tags", response.Tags)
+	d.Set("created_at", response.CreatedAt)
+	d.Set("updated_at", response.UpdatedAt)
+	d.Set("apiserver_url", response.ClusterURL)
+	d.Set("wildcard_dns", response.DNSWildcard)
+	d.Set("status", response.Status.String())
+
+	// autoscaler_config
+	autoscalerConfig := map[string]interface{}{}
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Move this in a `clusterAutoscalerConfigFlatten` helper func

File: scaleway/resource_k8s_cluster_beta.go [hunk 17]
```diff
@@ -0,0 +1,733 @@
...
+
+	d.Set("autoscaler_config", []map[string]interface{}{autoscalerConfig})
+
+	// default_pool_config
+	err = resourceScalewayK8SClusterBetaDefaultPoolRead(d, m)
+	if err != nil {
+		return err
+	}
+
+	////
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Move this in a `clusterKubeConfigFlatten` helper func

File: scaleway/resource_k8s_cluster_beta.go [hunk 18]
```diff
@@ -0,0 +1,733 @@
...
+}
+
+func autoscalerConfigSchema() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"disable_scale_down": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Computed:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set default value and remove computed

File: scaleway/resource_k8s_cluster_beta.go [hunk 19]
```diff
@@ -0,0 +1,733 @@
...
+				Description: "Disable the scale down feature of the autoscaler",
+			},
+			"scale_down_delay_after_add": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set default value and remove computed

File: scaleway/resource_k8s_cluster_beta.go [hunk 20]
```diff
@@ -0,0 +1,733 @@
...
+				Description: "",
+			},
+			"estimator": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set default value and remove computed

File: scaleway/resource_k8s_cluster_beta.go [hunk 21]
```diff
@@ -0,0 +1,733 @@
...
+				Description: "",
+			},
+			"expander": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set default value and remove computed

File: scaleway/resource_k8s_cluster_beta.go [hunk 22]
```diff
@@ -0,0 +1,733 @@
...
+				Description: "",
+			},
+			"ignore_daemonsets_utilization": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Computed:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set default value and remove computed

File: scaleway/resource_k8s_cluster_beta.go [hunk 23]
```diff
@@ -0,0 +1,733 @@
...
+				Description: "",
+			},
+			"balance_similar_node_groups": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Computed:    true,
+				Default:     nil,
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Set default value and remove computed

File: scaleway/helpers_k8s.go (hunk 4 above)
Comment (Sh4d1) [reply scaleway/helpers_k8s.go]: Hum, not sur I get the point

File: scaleway/resource_k8s_cluster_beta_test.go [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	k8s "github.com/scaleway/scaleway-sdk-go/api/k8s/v1beta3"
+)
+
+func TestAccScalewayK8SClusterBetaMinimal(t *testing.T) {
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_cluster_beta_test.go]: Can you test an update (add a step in your test)?

File: scaleway/resource_k8s_cluster_beta_test.go [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+		}
+
+		_, err = k8sAPI.GetCluster(&k8s.GetClusterRequest{
+			Region:    region,
+			ClusterID: clusterID,
+		})
+
+		// If no error resource still exist
+		if err == nil {
+			return fmt.Errorf("Cluster (%s) still exists", rs.Primary.ID)
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_cluster_beta_test.go]: cluster

File: scaleway/resource_k8s_pool_beta.go [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The name of the cluster",
+			},
+			"node_type": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Server Type of the pool servers",
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_pool_beta.go]: type

File: scaleway/resource_k8s_pool_beta.go [hunk 27]
```diff
@@ -0,0 +1,242 @@
...
+func resourceScalewayK8SPoolBetaRead(d *schema.ResourceData, m interface{}) error {
+	k8sAPI, region, poolID, err := getK8SAPIWithRegionAndID(m, d.Id())
+	if err != nil {
+		return err
+	}
+
+	////
+	// Read Pool
+	////
+
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_pool_beta.go]: remove line

File: scaleway/resource_k8s_pool_beta.go [hunk 28]
```diff
@@ -0,0 +1,242 @@
...
+func resourceScalewayK8SPoolBetaUpdate(d *schema.ResourceData, m interface{}) error {
+	k8sAPI, region, poolID, err := getK8SAPIWithRegionAndID(m, d.Id())
+	if err != nil {
+		return err
+	}
+
+	////
+	// Update Pool
+	////
+
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_pool_beta.go]: remove line

**Pull request merged**